src/deluge/util/pack.c

```diff
@@ -209,3 +208,45 @@ int32_t unpack_7to8_rle(uint8_t* dst, int32_t dst_size, uint8_t* src, int32_t sr
 	}
 	return d;
 }
+
+// derived from png reference implementation (MIT license)
+
+// Table of CRCs of all 8-bit messages.
+static uint32_t crc_table[256];
+
+/* Make the table for a fast CRC. */
+void init_crc_table(void) {
+	uint32_t c;
+	int n, k;
+
+	for (n = 0; n < 256; n++) {
+		c = (uint32_t)n;
+		for (k = 0; k < 8; k++) {
+			if (c & 1)
+				c = 0xedb88320L ^ (c >> 1);
+			else
+				c = c >> 1;
+		}
+		crc_table[n] = c;
+	}
+}
+
+// Update a running CRC with the bytes buf[0..len-1]--the CRC
+// should be initialized to all 1's, and the transmitted value
+// is the 1's complement of the final running CRC (see the
+// crc() routine below).
+
+static uint32_t update_crc(uint32_t crc, uint8_t* buf, int len) {
+	uint32_t c = crc;
+	int n;
+
+	for (n = 0; n < len; n++) {
+		c = crc_table[(c ^ buf[n]) & 0xff] ^ (c >> 8);
+	}
+	return c;
+}
+
+// Return the CRC of the bytes buf[0..len-1].
+uint32_t get_crc(uint8_t* buf, int len) {
```

Comment:
FatFS and `sd_util.c` already have CRC implementations, can we reuse one of those?

Reply:
maybe, but the idea is that this file `pack.c` is self-contained as it is meant to be used both by the deluge and client software. Also I could remember wrong, but those used the slow naive method instead of the table method so I think those could reuse this rather than the other way around.

Reply:
No problem, it's quite small.